lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -602,6 +602,143 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+/// Helper routine for preparing and propagating concrete type from
+/// protocol conformance analysis.
+bool SILCombiner::prepareNCheckSoleConformingTypePropagation(
+    FullApplySite Apply, ProtocolDecl *Protocol, SILValue &Arg,
+    ConcreteExistentialInfo &CEI) {
+
+  // Only applicable in whole-module compilation.
+  if (!Apply.getModule().isWholeModule())
+    return false;
+
+  // Determine the sole conforming type.
+  auto *NTD = PCA->findSoleConformingType(Protocol);
+  if (!NTD)
+    return false;
+
+  // Sole conforming class should not be public or have derived classes.
+  ClassDecl *CD;
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+      (CD->getEffectiveAccess() > AccessLevel::Internal ||
+       CHA->hasKnownDirectSubclasses(CD) ||
+       CHA->hasKnownIndirectSubclasses(CD))) {
+    return false;
+  }
+
+  // If the Arg is an AllocStackInst, determine the openexistential.
+  bool isCopied = false;
+  if (auto *Instance = dyn_cast<AllocStackInst>(Arg)) {
+    if (SILValue Src =
+            getAddressOfStackInit(Instance, Apply.getInstruction(), isCopied)) {
+      Arg = Src;
+    }
+  }
+  CEI.isCopied = isCopied;
+
+  if (!(isa<OpenExistentialRefInst>(Arg) || isa<OpenExistentialAddrInst>(Arg)))
+    return false;
+
+  auto ElementType = NTD->getDeclaredType();
+  CEI.ConcreteType = ElementType->getCanonicalType();
+  auto &M = Builder.getModule();
+  SILType ConcreteSILType = M.Types.getLoweredType(CEI.ConcreteType);
+  // We have the openexistential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(CEI.ConcreteType, Protocol);
+  if (!(ConformanceRef.hasValue()))
+    return false;
+
+  // Prepare the code by adding UncheckedCast instructions.
+  if (auto *OER = dyn_cast<OpenExistentialRefInst>(Arg)) {
+    auto *URCI =
+        Builder.createUncheckedRefCast(OER->getLoc(), OER, ConcreteSILType);
+    CEI.OpenedArchetype = OER->getType().getASTType()->castTo<ArchetypeType>();
+    CEI.ConcreteValue = URCI;
+  } else if (auto *OEA = dyn_cast<OpenExistentialAddrInst>(Arg)) {
+    auto *UACI = Builder.createUncheckedAddrCast(
+        OEA->getLoc(), OEA, ConcreteSILType.getAddressType());
+    CEI.OpenedArchetype = OEA->getType().getASTType()->castTo<ArchetypeType>();
+    CEI.ConcreteValue = UACI;
+  }
```

Comment:
```
else { llvm_unreacahble(...) }
```

Reply:
As discussed, will create a susbequent version for handling existential meta type.